source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc

```diff
@@ -47,33 +54,52 @@ const HeaderList AWSLambdaFilter::HeadersToSign =
          Http::Headers::get().ContentType});
 
 AWSLambdaFilter::AWSLambdaFilter(Upstream::ClusterManager &cluster_manager,
-                                 TimeSource &time_source)
-    : aws_authenticator_(time_source), cluster_manager_(cluster_manager) {}
+                                 TimeSource &time_source,
+                                 AWSLambdaConfigConstSharedPtr filter_config)
+    : aws_authenticator_(time_source), cluster_manager_(cluster_manager),
+      filter_config_(filter_config) {}
 
 AWSLambdaFilter::~AWSLambdaFilter() {}
 
-std::string AWSLambdaFilter::functionUrlPath(const std::string &name,
-                                             const std::string &qualifier) {
-
-  std::stringstream val;
-  val << "/2015-03-31/functions/" << name << "/invocations";
-  if (!qualifier.empty()) {
-    val << "?Qualifier=" << qualifier;
-  }
-  return val.str();
-}
-
 Http::FilterHeadersStatus
 AWSLambdaFilter::decodeHeaders(Http::HeaderMap &headers, bool end_stream) {
 
   protocol_options_ = Http::SoloFilterUtility::resolveProtocolOptions<
       const AWSLambdaProtocolExtensionConfig>(
       SoloHttpFilterNames::get().AwsLambda, decoder_callbacks_,
       cluster_manager_);
+
   if (!protocol_options_) {
     return Http::FilterHeadersStatus::Continue;
   }
 
+  const std::string *access_key{};
+  const std::string *secret_key{};
+  if (protocol_options_->accessKey().has_value() &&
+      protocol_options_->secretKey().has_value()) {
+    access_key = &protocol_options_->accessKey().value();
+    secret_key = &protocol_options_->secretKey().value();
+  } else if (filter_config_) {
+    credentials_ = filter_config_->getCredentials();
+    if (credentials_) {
+      const absl::optional<std::string> &maybeAccessKeyId =
+          credentials_->accessKeyId();
+      const absl::optional<std::string> &maybeSecretAccessKey =
+          credentials_->secretAccessKey();
+      if (maybeAccessKeyId.has_value() && maybeSecretAccessKey.has_value()) {
+        access_key = &maybeAccessKeyId.value();
+        secret_key = &maybeSecretAccessKey.value();
+      }
+    }
+  }
+
+  if ((access_key == nullptr) || (secret_key == nullptr)) {
+    decoder_callbacks_->sendLocalReply(
+        Http::Code::NotFound, RcDetails::get().CredentialsNotFoundBody, nullptr,
```

Comment:
Nit: is this really a 404: seems more like a 403 unauthorized

Reply:
i wouldn't say a 403; maybe 400 bad request?

Reply:
actually i think the best is 500 - internal server error